### DISCUSSION TITLE
Can I use AWS DocumentDB instead of MongoDB?

### DISCUSSION THREAD

**Discussion by prove-ability** [payloadcms/payload#652] [Q&A] [answered]
```mongodb://bard:<passwrd>@docdb-2022-06-13-02-18-07.cluster-c6wqayu9kx2t.ap-northeast-2.docdb.amazonaws.com:27017/?ssl=true&ssl_ca_certs=rds-combined-ca-bundle.pem&replicaSet=rs0&retryWrites=false```

I want to link AWS DocumentDB.
I want help.

<img width="792" alt="스크린샷 2022-06-13 오전 11 29 52" src="https://user-images.githubusercontent.com/59461181/173268912-3a7add16-0029-41d9-ae8e-fae17c3c9927.png">

**Comment by denolfe** [comment, marked as answer]
Hey @prove-ability, this error indicates that you are not passing the pem file contents into the Mongo options. You will need to have your .pem file locally and pass the contents into `payload.init` under `mongoOptions`. Here is an (untested) example:

```ts
const caContent = fs.readFileSync('/path/to/rds-combined-ca-bundle.pem');

payload.init({
  // ..
  mongoOptions: {
    sslCA: caContent,
  },
});
```

Let me know if that gets you any further.

Here are some links to SO issues that may be useful as well: [link](https://stackoverflow.com/questions/54397644/connection-error-while-connecting-to-aws-documentdb), [link](https://stackoverflow.com/questions/59570957/unable-to-determine-mongoose-ssl-connection-error-to-documentdb)

**Reply by prove-ability** [reply]
The method that I told you didn't work.

So I tried to proceed after watching [AWS DOCS](https://docs.aws.amazon.com/documentdb/latest/developerguide/connect_programmatically.html), but this also doesn't work.

Do you have any idea?

```javascript
// Initialize Payload
payload.init({

  secret: process.env.PAYLOAD_SECRET,
  mongoURL: process.env.MONGODB_URI,
  express: app,
  onInit: () => {
    payload.logger.info(`Payload Admin URL: ${payload.getAdminURL()}`)
  },
  mongoOptions: {
    tlsCAFile: `${__dirname}/rds-combined-ca-bundle.pem`
    // ssl: true,
    // sslValidate: false,
    // sslCA: `${__dirname}/rds-combined-ca-bundle.pem`
  }
})
```

**Reply by x31b** [reply]
hello @prove-ability 
  
Besides the SSL-issue, do you think DocumentDB's MongoDB API compatibility sufficient for PayloadCMS purposes?

**Reply by jmikrut** [reply]
@x31b — yes, you should be able to 100% use DocumentDB!

**Reply by x31b** [reply]
Awesome. Thanks @jmikrut

**Reply by denolfe** [reply]
Also worth looking at @hdodov 's comment below as well: https://github.com/payloadcms/payload/discussions/652#discussioncomment-6291517

**Comment by moltar** [comment]
@prove-ability Have you gotten the DocumentDB to work?

**Comment by PukpikC** [comment]
Hello @jmikrut ,
I setup payload on aws but got this error when add data "You are not allowed to perform this action"
![image](https://user-images.githubusercontent.com/24909156/231764710-497c6103-898b-41e1-bbf3-85067eb25c7f.png)

i tried set `mongoOptions` up as in suggestions above but it still not working

related error
![image](https://user-images.githubusercontent.com/24909156/231795665-b539782a-357b-4830-840b-904414cb37a6.png)


Do you have Ideas why, 
Thank you ^^

**Reply by DanRibbens** [reply]
If you're getting as far as you are in these screenshots, like being able to login, then your issue isn't the database configuration. It has to be your access control or a cors issue. To verify, do you still get the same errors when using MongoDB instead of DocumentDB? I'm assuming that you would.

@PukpikC I would open a new Q&A discussion if you want troubleshooting help and include some of your config and steps to reproduce.

**Reply by PukpikC** [reply]
@DanRibbens Thank you for your answer, ah yes everything works fine on my local using mongoDB. and it would be really nice for new Q&A discussion. 😃

**Reply by DanRibbens** [reply]
I can't understand how only changing the database connection would make this happen. Your other requests would be failing before these if it was the DB. I still think there is something else going on.

**Reply by PukpikC** [reply]
@DanRibbens Hi, Than you for helping, I found a problem, it was about access control. I have to set it correctly to make it work on live :D

**Comment by yandodov** [comment]
I'm going to chime in with how I managed to connect to AWS DocumentDB locally.

As stated in [the AWS docs about "Connecting to an Amazon DocumentDB Cluster from Outside an Amazon VPC"](https://docs.aws.amazon.com/documentdb/latest/developerguide/connect-from-outside-a-vpc.html), you need to use SSH tunneling (port forwarding). Here's the command that creates a tunnel:

```bash
ssh -i /Users/path/to/ssh-private-key -L 27017:your-cluster.cluster-id.eu-central-1.docdb.amazonaws.com:27017 ec2-user@ec2-12-345-678-90.eu-central-1.compute.amazonaws.com -N
```

Then, your `MONGODB_URI` connection string must look like:

```
mongodb://dbuser:dbpassword@localhost:27017/?tls=true&tlsCAFile=eu-central-1-bundle.pem&tlsInsecure=true&directConnection=true&retryWrites=false
```

Couple of things to note:

- `eu-central-1-bundle.pem` can be downloaded from [the AWS docs on "Using SSL/TLS to encrypt a connection to a DB instance"](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html#UsingWithRDS.SSL.RegionCertificates) and put in the root directory of the project.

- `tlsInsecure=true` is needed because due to the SSH tunneling, you're actually connecting to `localhost` and you would receive the following error regarding the TLS certificate:

  > Error: cannot connect to MongoDB. Details: Hostname/IP does not match certificate's altnames: Host: localhost. is not in the cert's altnames: DNS:your-cluster.cluster-id.eu-central-1.docdb.amazonaws.com, DNS:your-cluster.cluster-cluster-id.eu-central-1.docdb.amazonaws.com, DNS:your-cluster.cluster-ro-cluster-id.eu-central-1.docdb.amazonaws.com

- `directConnection=true` must be added, as mentioned in [this Mongoose issue](https://github.com/Automattic/mongoose/issues/11105#issuecomment-1048589468). Quoting [the MongoDB Go Driver docs](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.12.0/mongo/options#ClientOptions):

  > If set to true, the driver will only connect to the host provided in the URI and will not discover other hosts in the cluster.

  My guess is that this is needed because otherwise the SSH tunnel is messing with that "host discovery".
- `retryWrites=false` is needed because otherwise you get:

  > MongoServerError: Retryable writes are not supported

At this point, you should have `localhost:27017` forwarding the connection to `your-cluster.cluster-id.eu-central-1.docdb.amazonaws.com:27017` using a tunnel through your EC2 instance `ec2-12-345-678-90.eu-central-1.compute.amazonaws.com` via `ec2-user` after authenticating with your EC2 SSH private key `/Users/path/to/ssh-private-key`.

---

Before using DocumentDB, I was using Atlas, and the only change in my code I had to make (besides the MongoDB connection string in `.env`) was to disable [the `useFacet` setting](https://payloadcms.com/docs/production/deployment#mongodb):

```diff
await payload.init({
	secret: process.env.PAYLOAD_SECRET,
	mongoURL: process.env.MONGODB_URI,
	express: app,
+	mongoOptions: {
+		useFacet: false,
+	},
});
```

**Comment by DmitriyNoa** [comment]
Hello. I was able to connect to DocumentDb with following config
`mongooseAdapter({
    url: process.env.DATABASE_URI || '',
    disableIndexHints: true,
    transactionOptions: false,
    connectOptions: {
      timeoutMS: 1000,
      ssl: true,
      directConnection: true,
      tlsInsecure: true,
      retryWrites: false,
      useFacet: false,
    }
  })`
The basic CMS UI works. I'm able to get the documents and edit. However as soon as I try to use where queries all requests are stop working and basically hangs until the timeout. Even the default filtration in the UI doesn't work.

Does anyone know how to fix this issue?

P.S. I have my own custom endpoint that is using payload.db. When I use .sort() the query stops working and showing the same issue. But if removed it works as expected.
